### PR TITLE
Add some basic proxy protocol tests

### PR DESCRIPTION
**Pull request by bleggett**
[PROXY protocol support in Istio is currently marked as experimental](https://istio.io/latest/docs/ops/configuration/traffic-management/network-topologies/#proxy-protocol) and lacks any integration tests - this PR adds a few simple integration tests to the existing gateway topology suite in an effort to move it forward to Alpha status.

Sibling PR for doc update: https://github.com/istio/istio.io/pull/12510

- Really, Istio isn't very involved with PROXY protocol support - it's an Envoy config.
- PROXY protocol support is realistically only for proxying TCP traffic from a L4 loadbalancer thru a Gateway to a backend service - using PROXY protocol for L7 traffic is a bad idea, as this sibiling doc update attempts to make clear.
- Currently only PROXY protocol  V1 support is tested - this is all the major cloud L4s support ATM anyway.
- Pulls in `github.com/pires/go-proxyproto` as an integration test rig dependency - this lib is reasonably well tested and maintained and using it seemed better than implementing a L4 header library from scratch in the test rig.

### CODE REVIEW

File: tests/integration/pilot/common/traffic.go [hunk 1]
```diff
@@ -210,7 +211,7 @@ func (c TrafficTestCase) Run(t framework.TestContext, namespace string) {
 		if len(c.config) > 0 {
 			cfg := yml.MustApplyNamespace(t, c.config, namespace)
 			// we only apply to config clusters
-			t.ConfigIstio().YAML("", cfg).ApplyOrFail(t)
+			t.ConfigIstio().YAML("", cfg).ApplyOrFail(t, apply.NoCleanup)
```
Comment (howardjohn) [inline tests/integration/pilot/common/traffic.go:214]: was this for debugging on intended to merge?

File: tests/integration/pilot/gw_topology_test.go [hunk 2]
```diff
@@ -90,7 +91,7 @@ spec:
         image: auto
         imagePullPolicy: {{ .imagePullPolicy }}
 ---
-`).ApplyOrFail(t)
+`).ApplyOrFail(t, apply.NoCleanup)
```
Comment (howardjohn) [inline tests/integration/pilot/gw_topology_test.go:94]: same comment

**Review by howardjohn — COMMENTED** [review]
skimmed 75% of it, seems reasonable. will finish review tomorrow. Thanks!

**Comment by kyessenov** [comment]
Is there a corresponding API proposal? We can't support EnvoyFilter long term.

File: tests/integration/pilot/common/traffic.go (hunk 1 above)
Comment (bleggett) [reply tests/integration/pilot/common/traffic.go:214]: Should be `CleanupConditionally` - was having issues with the test suite nuking stuff out from under me.

File: tests/integration/pilot/gw_topology_test.go (hunk 2 above)
Comment (bleggett) [reply tests/integration/pilot/gw_topology_test.go:94]: Yep, fixed.

**Comment by bleggett** [comment]
> Is there a corresponding API proposal? We can't support EnvoyFilter long term.

1. No, and agree - context: Several PRs have cropped up around people asking how to do this, or getting confused about what it's for, and currently all/most cloud provider L4 load balancers support PROXY protocol, so in the [sibling doc PR](https://github.com/istio/istio.io/pull/12510) I've attempted to contextualize the usage a bit better.

**Comment by kyessenov** [comment]
> > Is there a corresponding API proposal? We can't support EnvoyFilter long term.
> 
> No, and agree - context: Several PRs have cropped up around people asking how to do this, or getting confused about what it's for, and currently all/most cloud provider L4 load balancers support PROXY protocol, so in the [sibling doc PR](https://github.com/istio/istio.io/pull/12510) I've attempted to contextualize the usage a bit better.
> 
> I think a few basic tests and a more clear doc pointer indicating where to go in the Envoy docs to do this (such as we have for other gateway config options, as I added in the sibling PR) are enough for now.
> 
> Other thoughts welcome.

I completely agree that the tests and the guidance are useful. But I also don't want this work to end up like the local rate limiting in Istio. We need a path to API, using the gateway API ideally.

**Comment by costinm** [comment]
I'm not entirely sure it needs an API (in the sense of a CR ) - it is more
aligned with auto-mtls or ZTunnel support, which is also based on the PROXY
support.

The only difference between zTunnel and a 'legacy' proxy is that zTunnel
requires H2 and an outer encryption - but Squid and other proxies are
starting to
add this as well ( since Firefox/Chrome have added it too ).

Also this is an infrastructure property - not something a regular app
developer would know about or be able to configure.

I suggest continuing to use labels, as is current practice for auto-mtls
and ztunnel - we may attempt to standardize something upstream, but
we need to be careful. Maybe as part of Service v2 - appProtocol is already
problematic.

**Comment by kyessenov** [comment]
@costinm I think this is a different feature - this is for ALB/cloud LBs providing the original source for XFF and policy. But my point stands - we need a good feature definition (via API or whatever), and not a bunch of EnvoyFilter scripts for ChatGPT :)

**Comment by costinm** [comment]
Sorry, my mistake - I thought HTTP PROXY, not the HA PROXY.

But like ChatGPT - I may still be right even if I had no idea what I am
talking about :-) HAProxy is also an 'infrastructure specific' and not
something
the app developer controls.

We do have an API for 'proxies in front of the gateway' in MeshConfig, and
it can be extended in a few ways - the biggest problem
right now is that it doesn't define any restrictions on who is allowed to
talk with the gateway (CIDR, etc) nor does it document
very well that NetworkPolicy or other means need to be used to make sure
only front gateways can connect.

**Comment by linsun** [comment]
> > > Is there a corresponding API proposal? We can't support EnvoyFilter long term.
> > 
> > 
> > No, and agree - context: Several PRs have cropped up around people asking how to do this, or getting confused about what it's for, and currently all/most cloud provider L4 load balancers support PROXY protocol, so in the [sibling doc PR](https://github.com/istio/istio.io/pull/12510) I've attempted to contextualize the usage a bit better.
> > I think a few basic tests and a more clear doc pointer indicating where to go in the Envoy docs to do this (such as we have for other gateway config options, as I added in the sibling PR) are enough for now.
> > Other thoughts welcome.
> 
> I completely agree that the tests and the guidance are useful. But I also don't want this work to end up like the local rate limiting in Istio. We need a path to API, using the gateway API ideally.

This is different than rate limiting as people are complaining the docs state it is experimental. I completely agree with you that in order to promote it beyond alpha we would need to define a proper API other than EnvoyFilter, but I don't think marking it as experimental while people wants to use it and some users are already running it in production makes sense. Ben's PR gives us the confidence that it is properly tested for alpha status.

**Comment by howardjohn** [comment]
IMO there is no feature status. We have a feature - Envoyfitler - which is permanently alpha and unstable. We have examples of using it (https://github.com/istio/istio/wiki/EnvoyFilter-Samples); Proxy protocol is one example. Its reasonable to have tests for it. We should not call "proxy protocol" a feature of Istio with a stability gaurantee

**Comment by howardjohn** [comment]
Forgot to mention - having a test is still useful regardless

File: tests/integration/pilot/gw_topology_test.go [hunk 3]
```diff
@@ -101,3 +102,175 @@ spec:
 			}
 		})
 }
+
+// If the PROXY protocol is enabled, and no XFF header is provided, client forwarded IPs from the PROXY protocol payload
+// should be used to populate the upstream XFF header.
+//
+// If the PROXY protocol is enabled, and additionally an XFF header is provided AND numTrustedProxies is set, the XFF header on the incoming request
+// will take precedence when populating the upstream XFF header.
+func TestProxyProtocolGatewayEnabled(t *testing.T) {
```
Comment (howardjohn) [inline tests/integration/pilot/gw_topology_test.go:111]: Its best to have 1 deployment of the GW and have many subtests. Can we merge this and the one below, or maybe even all 3 into TestXFFGateway?

Deploying a new GW deployment takes ~20s. Deploying a new EnvoyFilter for it takes 100ms. So where possible we should avoid new deployments

Comment (bleggett) [reply tests/integration/pilot/gw_topology_test.go:111]: Considered + initially tried that and felt it resulted in a less obvious set of tests - that creates an implicit ordering for test cases tho (test without filter, apply filter as last test) - rigs don't support "removing" resources except via teardown AFAICT.

Will rework.

Comment (howardjohn) [reply tests/integration/pilot/gw_topology_test.go:111]: You can tear down. This is how 99% of our tests work (see TestTraffic, its one test with like 100 sub tests). You just need to use a subtest.


```
Run(
  common setup
  Subtest(
   test something...
  )
  Subtest(
   test something...
  )
  Subtest(
   more setep...
     SubTest(usesetup)
  )
```

Comment (bleggett) [reply tests/integration/pilot/gw_topology_test.go]: Done - 1 less gateway created during test run.

**Comment by bleggett** [comment]
It's unclear to me why the `arm64` suites are failing specifically.

**Comment by howardjohn** [comment]
had an outage on arm64 cluster, fixed now

/retest

**Comment by bleggett** [comment]
Because of a botched initial diff several other people got sucked into this as reviewers - apologies.

File: tests/integration/pilot/common/routing.go [hunk 4]
```diff
@@ -1570,8 +1604,13 @@ spec:
 	}
 }
 
-func XFFGatewayCase(apps *deployment.SingleNamespaceView, gateway string) []TrafficTestCase {
+// 1. Creates a TCP Gateway and VirtualService listener
+// 2. Configures the echoserver to call itself via the TCP gateway using PROXY protocol https://www.haproxy.org/download/1.8/doc/proxy-protocol.txt
+// 3. Assumes that the proxy filter EnvoyFilter not is applied
```
Comment (linsun) [inline tests/integration/pilot/common/routing.go:1609]:
```suggestion
// 3. Assumes that the proxy filter EnvoyFilter is not applied
```

**Review by linsun — APPROVED** [review]
LGTM, thanks @bleggett for addressing all comments from various reviewers.

**Review by howardjohn — APPROVED** [review]

**Pull request merged**